events/events.go

```diff
@@ -0,0 +1,138 @@
+package events
+
+import (
+	"context"
+	"time"
+
+	cs "github.com/cloudtrust/common-service/v2"
+	"github.com/cloudtrust/common-service/v2/events/fb"
+	"github.com/cloudtrust/common-service/v2/log"
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/google/uuid"
+)
+
+const (
+	CtEventType            = "ct_event_type"
+	CtEventAgentUsername   = "agent_username"
+	CtEventAgentRealmName  = "agent_realm_name"
+	CtEventTargetUserID    = "target_user_id"
+	CtEventGroupID         = "group_id"
+	CtEventGroupName       = "group_name"
+	CtEventRoleID          = "role_id"
+	CtEventRoleName        = "role_name"
+	CtEventOrigin          = "origin"
+	CtEventAuditTime       = "audit_time"
+	CtEventTargetRealmName = "target_realm_name"
+	CtEventAgentUserID     = "agent_user_id"
+	CtEventTargetUsername  = "target_username"
+	CtEventKcEventType     = "kc_event_type"
+	CtEventKcOperationType = "kc_operation_type"
+	CtEventClientID        = "client_id"
+	CtEventAdditionalInfo  = "additional_info"
+
+	CtEventUnknownUsername = "--UNKNOWN--"
+)
+
+type Event struct {
+	time      time.Time
+	origin    string
+	eventType string
+	details   map[string]string
+}
+
+func newEvent(origin string, eventType string, agentRealmName string, agentUserID string, agentUsername string, targetRealmName string, targetUserID *string, targetUserName *string, details map[string]string) Event {
+	if details == nil {
+		details = map[string]string{}
+	}
+
+	details[CtEventAgentRealmName] = agentRealmName
+	details[CtEventAgentUserID] = agentUserID
+	details[CtEventAgentUsername] = agentUsername
+	details[CtEventTargetRealmName] = targetRealmName
+	if targetUserID != nil {
+		details[CtEventTargetUserID] = *targetUserID
+	}
+	if targetUserName != nil {
+		details[CtEventTargetUsername] = *targetUserName
+	}
+
+	return Event{
+		time:      time.Now().UTC(),
+		origin:    origin,
+		eventType: eventType,
+		details:   details,
+	}
+}
+
+func NewEventOnUser(origin string, eventType string, agentRealmName string, agentUserID string, agentUsername string, targetRealmName string, targetUserID string, targetUserName string, details map[string]string) Event {
+	return newEvent(origin, eventType, agentRealmName, agentUserID, agentUsername, targetRealmName, &targetUserID, &targetUserName, details)
+}
+
+func NewEventOnUserFromContext(ctx context.Context, logger log.Logger, origin string, eventType string, targetRealmName string, targetUserID string, targetUserName string, details map[string]string) Event {
```

Comment:
NewEventOnAgentFromContext? how can the user be sure that the user in the context is considered as the user agent?

Reply:
The agent will be extracted from the context.
The target user will be taken from the params. 

NewEventOnAgentFromContext is not clearer to me and is not very meaningful